### PR TITLE
#2455 Replace boost::unordered_set by std::unordered_set

### PR DESCRIPTION
**Pull request by barondim**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [ ] unit tests and non-regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [x] the corresponding milestone was added in the ticket and in this PR
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated

### CODE REVIEW

**Comment by barondim** [comment]
Only a part of the boost containers were replaced by std containers to decrease the task complexity.

File: dynawo/sources/API/PAR/test/TestParametersSet.cpp [hunk 1]
```diff
@@ -215,8 +215,7 @@ TEST(APIPARTest, ParametersSetGetReferences) {
     ++nbReferences;
   ASSERT_EQ(nbReferences, 3);
 
-  ParametersSet::reference_const_iterator itRef = parametersSet->cbeginReference();
-  ASSERT_EQ(itRef->get()->getName(), "ref3");
+  ASSERT_EQ(ref3->getName(), "ref3");
```
Comment (rosiereflo) [inline dynawo/sources/API/PAR/test/TestParametersSet.cpp:218]: Please do not modify the unit tests. This one was made to test cbeginReference

File: dynawo/sources/Common/DYNNumericalUtils.h [hunk 2]
```diff
@@ -14,6 +14,8 @@
 #ifndef COMMON_DYNNUMERICALUTILS_H_
 #define COMMON_DYNNUMERICALUTILS_H_
 
+#include <cmath>
```
Comment (rosiereflo) [inline dynawo/sources/Common/DYNNumericalUtils.h:17]: Should be after DYNMacrosMessage.h

File: dynawo/sources/Modeler/Common/DYNCompiler.cpp [hunk 3]
```diff
@@ -74,8 +73,6 @@ using std::vector;
 using boost::dynamic_pointer_cast;
 using boost::lexical_cast;
 using boost::shared_ptr;
-using boost::unordered_map;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNCompiler.cpp:77]: Why removing the using instead of modifying them?

File: dynawo/sources/Modeler/Common/DYNDelayManager.h [hunk 4]
```diff
@@ -20,10 +20,11 @@
 #ifndef MODELER_COMMON_DYNDELAYMANAGER_H_
 #define MODELER_COMMON_DYNDELAYMANAGER_H_
 
+#include <unordered_map>
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNDelayManager.h:23]: Should be left where it was

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 5]
```diff
@@ -151,6 +150,26 @@ using parameters::ParametersSetCollection;
 
 static const char TIME_FILENAME[] = "time.bin";  ///< name of the file to dump time at the end of the simulation
 
+
+/**
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.cpp:154]: Why is this required?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/API/PAR/test/TestParametersSet.cpp (hunk 1 above)
Comment (barondim) [reply dynawo/sources/API/PAR/test/TestParametersSet.cpp:218]: There are a few things I don't understand :
cbeginReference() is called elsewhere in the unit test so it is actually tested even if I delete this line
I don't understand what "begin" means here. An unordered_map has actually no beginning. I had to change this unit test because the std version worked locally but had a different result on the CI.

File: dynawo/sources/Modeler/Common/DYNCompiler.cpp (hunk 3 above)
Comment (barondim) [reply dynawo/sources/Modeler/Common/DYNCompiler.cpp:77]: I feel that the "using" keyword makes the code less clear

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 5 above)
Comment (barondim) [reply dynawo/sources/Simulation/DYNSimulation.cpp:154]: The key of unordered_map<std::pair<std::string, std::string>, boost::shared_ptr<curves::Curve> > is a std::pair<std::string, std::string> and the Hash is not already defined so I had to define it.

File: dynawo/sources/API/PAR/test/TestParametersSet.cpp (hunk 1 above)
Comment (rosiereflo) [reply dynawo/sources/API/PAR/test/TestParametersSet.cpp:218]: parametersSet is not a unordered_map

File: dynawo/sources/Modeler/Common/DYNCompiler.cpp (hunk 3 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNCompiler.cpp:77]: This is creating a lot of differences that will make harder to explore the history of this file, I'm not fond of it

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 5 above)
Comment (rosiereflo) [reply dynawo/sources/Simulation/DYNSimulation.cpp:154]: This was already true earlier right?

File: dynawo/sources/Modeler/Common/DYNCompiler.cpp (hunk 3 above)
Comment (barondim) [reply dynawo/sources/Modeler/Common/DYNCompiler.cpp:77]: I put back the using in that case

File: dynawo/sources/Simulation/DYNSimulation.cpp (hunk 5 above)
Comment (barondim) [reply dynawo/sources/Simulation/DYNSimulation.cpp:154]: It seems we didn't have to define it when we use boost::unordered_map

File: dynawo/sources/Modeler/Common/DYNCompiler.cpp (hunk 3 above)
Comment (barondim) [reply dynawo/sources/Modeler/Common/DYNCompiler.cpp:77]: I've just realized the using keyword can create confusion in the files which contains both boost containers and std containers. I'm not sure it's a good idea to restore these changes.

**Comment by barondim** [comment]
To be merged with https://github.com/dynawo/dynaflow-launcher/pull/731

**Comment by barondim** [comment]
To be merged after https://github.com/dynawo/dynawo/pull/3154

File: dynawo/sources/Common/DYNTrace.cpp [hunk 6]
```diff
@@ -198,7 +198,12 @@ void Trace::addAppenders_(const std::vector<TraceAppender>& appenders) {
 
   {
     boost::lock_guard<boost::mutex> lock(mutex_);
-    sinks_.insert_or_assign(currentId, traceSink);
+
```
Comment (rosiereflo) [inline dynawo/sources/Common/DYNTrace.cpp:201]: WHy modifying this? It seems that insert_or_assign method is also defined in std

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Common/DYNTrace.cpp (hunk 6 above)
Comment (barondim) [reply dynawo/sources/Common/DYNTrace.cpp:201]: insert_or_assign() is only available since C++17
